File: 1_dfs_bfs/강대훈/피로도.swift
import Foundation

func solution(_ k: Int, _ dungeons: [[Int]]) -> Int {
    var ans = 0
    var visited = [Bool](repeating: false, count: dungeons.count)

    dfs(k, 0, 0)

    func dfs(_ hp: Int, _ idx: Int, _ cnt: Int) {
        ans = max(ans, cnt)

        if idx >= dungeons.count {
            return
        }

        // 1 2 3
        // 1 3 2
        // 2 1 3
        // 2 3 1
        // 3 1 2
        // 3 2 1

        for i in 0..<dungeons.count {
            if visited[i] { continue }
            if hp < dungeons[i][0] { continue }

            visited[i] = true
            dfs(hp - dungeons[i][1], idx + 1, cnt + 1)
            visited[i] = false
        }
    }

    return ans
}

File: 3_bruteforce_backtracking/김민주/피로도.java

class Solution {
    int n;
    int mx =0;
    boolean[] visited;
    public int solution(int k, int[][] dungeons) {
        // sort 할지 말지 생각해보기
        n = dungeons.length;
        visited = new boolean[n];
        dfs(0, k, dungeons);
        
        return mx;
    }
    
    void dfs(int ans, int hp, int[][] dg){
        mx = Math.max(ans, mx);
        for (int i=0; i<n; i++){
            if (visited[i]){ continue; }
            
            if (hp >= dg[i][0]){
                visited[i] = true;
                dfs(ans+1, hp- dg[i][1], dg);
                
                visited[i] = false;
            }
            
            
        }
        return;
    }
}